src/metabase/models/database.clj

```diff
@@ -155,9 +155,32 @@
   "**MetabasePass**")
 
 (def ^:const sensitive-fields
-  "List of fields that should be obfuscated in API responses, as they contain sensitive data."
-  [:password :pass :tunnel-pass :tunnel-private-key :tunnel-private-key-passphrase
-   :access-token :refresh-token :service-account-json])
+  "Set of fields that should be obfuscated in API responses, as they contain sensitive data."
+  #{:password :pass :tunnel-pass :tunnel-private-key :tunnel-private-key-passphrase
+    :access-token :refresh-token :service-account-json})
+
+(defn get-sensitive-fields
+  "Gets all sensitive fields that should be redacted in API responses. This includes everything from the
+   sensitive-fields set, along with any custom properties from connection-properties which are marked with a
+   :type :password."
+  [driver]
+  (if
+    (contains? (driver.u/available-drivers) driver)
+    ;; we have a valid driver; check its connection-properties for any password fields
+    (let [all-fields      (driver/connection-properties driver)
+          password-fields (filter #(= (get % :type) :password) all-fields)]
+      (into  sensitive-fields (map (comp keyword :name) password-fields)))
+    ;; no valid driver, just return the default sensitive-fields
+    sensitive-fields))
+
+(defn get-sensitive-fields-for-db
+  "Gets all sensitive fields that should be redacted in API responses for a given database. Calls get-sensitive-fields
+  using the given database's driver, if that driver is valid and registered. Refer to get-sensitive-fields Scaladoc
+  for full details."
+  [database]
+  (if (empty? database)
+    (get-sensitive-fields nil) ;; preserve existing behavior by returning the default sensitive fields
+    (get-sensitive-fields (driver.u/database->driver database))))
```

Comment:
Is this better suited as a `(defmulti)` that can be implemented in each driver? See something like: https://github.com/metabase/metabase/blob/master/src/metabase/driver.clj#L416-L432

Reply:
I don't think this is something we'd want to use `defmulti` for -- this is just derived directly from `metabase.driver/connection-properties` which is already a multimethod. I don't see a situation where we'd have a different list of sensitive fields in `connection-properties` vs what comes back here

Reply:
Have a look at the refactor.